docs/site/Interceptors.md

```diff
@@ -0,0 +1,328 @@
+  constructor(@inject('valid-names') private validNames: string[]) {}
+  value() {
+    return async <T>(
+      invocationCtx: InvocationContext,
+      next: () => ValueOrPromise<T>,
+    ) => {
+      const name = invocationCtx.args[0];
+      if (!this.validNames.includes(name)) {
+        throw new Error(
+          `Name '${name}' is not on the list of '${this.validNames}`,
+        );
+      }
+      return await next();
+    };
+  }
+}
+```
+
+## Apply interceptors
+
+Interceptors form a cascading chain of handlers around the target method
+invocation. We can apply interceptors by decorating methods/classes with
+`@intercept`.
+
+### @intercept
+
+Syntax: `@intercept(...interceptorFunctionsOrBindingKeys)`
+
+The `@intercept` decorator adds interceptors to a class or its methods including
+static and instance methods. Two flavors are accepted:
+
+- An interceptor function
+
+  ```ts
+  class MyController {
+    @intercept(log) // Use the `log` function
+    greet(name: string) {
+      return `Hello, ${name}`;
+    }
+  }
+  ```
+
+- A binding key that can be resolved to an interface function
+
+  ```ts
+  class MyController {
+    @intercept('name-validator') // Use the `name-validator` binding
+    async helloWithNameValidation(name: string) {
+      return `Hello, ${name}`;
+    }
+  }
+
+  // Bind options and provider for `NameValidator`
+  ctx.bind('valid-names').to(['John', 'Mary']);
+  ctx.bind('name-validator').toProvider(NameValidator);
+  ```
+
+### Method level interceptors
+
+A static or prototype method on a class can be decorated with `@intercept` to
+apply interceptors. For example,
+
+```ts
+class MyController {
+  @intercept(log)
+  static async greetStatic(name: string) {
+    return `Hello, ${name}`;
+  }
+
+  @intercept(log)
+  static async greetStaticWithDI(@inject('name') name: string) {
+    return `Hello, ${name}`;
+  }
+
+  @intercept(logSync)
+  greetSync(name: string) {
+    return `Hello, ${name}`;
+  }
+
+  @intercept(log)
+  greet(name: string) {
+    return `Hello, ${name}`;
+  }
+
+  @intercept('log')
+  async hello(name: string) {
+    return `Hello, ${name}`;
+  }
+
+  @intercept(log)
+  async greetWithDI(@inject('name') name: string) {
+    return `Hello, ${name}`;
+  }
+
+  @intercept('log', logSync)
+  async helloWithTwoInterceptors(name: string) {
+    return `Hello, ${name}`;
+  }
+
+  async helloWithoutInterceptors(name: string) {
+    return `Hello, ${name}`;
+  }
+
+  @intercept(changeName)
+  async helloWithChangeName(name: string) {
+    return `Hello, ${name}`;
+  }
+
+  @intercept(logWithErrorHandling)
+  async helloWithError(name: string) {
+    throw new Error('error: ' + name);
+  }
+}
+```
+
+### Class level interceptors
+
+To apply interceptors to be invoked for all methods on a class, we can use
+`@intercept` to decorate the class. When a method is invoked, class level
+interceptors (if not explicitly listed at method level) are invoked before
+method level ones.
```

Comment:
Please clarify whether the interceptors will be applied to protected and private methods too.

Reply:
> Please clarify whether the interceptors will be applied to protected and private methods too.

I don't see where is this question answered, could you please point me to the right place?

Reply:
See https://github.com/strongloop/loopback-next/blob/interceptor/docs/site/Interceptors.md#method-level-interceptors